File: backend/pkg/database/repository/page.go
package repository

import (
	"context"
	"encoding/json"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// PageRepository defines the interface for page data access
type PageRepository interface {
	FindByID(ctx context.Context, id uuid.UUID) (*types.Page, error)
	FindBySlug(ctx context.Context, slug string) (*types.Page, error)
	List(ctx context.Context, opts types.PageListOptions) ([]types.Page, int64, error)
	Save(ctx context.Context, page *types.Page) error
	Delete(ctx context.Context, id uuid.UUID) error
}

// pageRepository provides data access for pages
type pageRepository struct {
	db *gorm.DB
}

// NewPageRepository creates a new PageRepository
func NewPageRepository(db *gorm.DB) PageRepository {
	return &pageRepository{db: db}
}

// pageModelToType converts a database model to types
func pageModelToType(m *models.Page) *types.Page {
	var metaData map[string]interface{}
	if m.MetaData != nil {
		_ = json.Unmarshal(m.MetaData, &metaData)
	}

	return &types.Page{
		ID:          m.ID,
		Slug:        m.Slug,
		Title:       m.Title,
		Content:     m.Content,
		Description: m.Description,
		ImageURL:    m.ImageURL,
		MetaData:    metaData,
		IsPublished: m.IsPublished,
		CreatedAt:   m.CreatedAt,
		UpdatedAt:   m.UpdatedAt,
	}
}

// pageTypeToModel converts a types type to database model
func pageTypeToModel(p *types.Page) *models.Page {
	var metaData datatypes.JSON
	if p.MetaData != nil {
		data, _ := json.Marshal(p.MetaData)
		metaData = datatypes.JSON(data)
	}

	return &models.Page{
		ID:          p.ID,
		Slug:        p.Slug,
		Title:       p.Title,
		Content:     p.Content,
		Description: p.Description,
		ImageURL:    p.ImageURL,
		MetaData:    metaData,
		IsPublished: p.IsPublished,
		CreatedAt:   p.CreatedAt,
		UpdatedAt:   p.UpdatedAt,
	}
}

// FindByID retrieves a page by its ID
func (r *pageRepository) FindByID(ctx context.Context, id uuid.UUID) (*types.Page, error) {
	var model models.Page
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return pageModelToType(&model), nil
}

// FindBySlug retrieves a page by its slug
func (r *pageRepository) FindBySlug(ctx context.Context, slug string) (*types.Page, error) {
	var model models.Page
	if err := r.db.WithContext(ctx).Where("slug = ?", slug).First(&model).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return pageModelToType(&model), nil
}

// List retrieves pages with pagination
func (r *pageRepository) List(ctx context.Context, opts types.PageListOptions) ([]types.Page, int64, error) {
	var pageModels []models.Page
	var total int64

	query := r.db.WithContext(ctx).Model(&models.Page{})

	// Apply IsPublished filter if specified
	if opts.IsPublished != nil {
		query = query.Where("is_published = ?", *opts.IsPublished)
	}

	// Get total count
	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	// Apply pagination
	offset := (opts.Page - 1) * opts.PerPage
	if err := query.Offset(offset).Limit(opts.PerPage).Order("created_at DESC").Find(&pageModels).Error; err != nil {
		return nil, 0, err
	}

	pages := make([]types.Page, len(pageModels))
	for i, m := range pageModels {
		pages[i] = *pageModelToType(&m)
	}

	return pages, total, nil
}

// Save creates or updates a page
func (r *pageRepository) Save(ctx context.Context, page *types.Page) error {
	model := pageTypeToModel(page)

	// Check if page exists
	var existing models.Page
	err := r.db.WithContext(ctx).First(&existing, model.ID).Error
	if err == gorm.ErrRecordNotFound {
		// Create new page
		if model.ID == uuid.Nil {
			model.ID = uuid.New()
			page.ID = model.ID
		}
		return r.db.WithContext(ctx).Create(model).Error
	} else if err != nil {
		return err
	}

	// Update existing page
	return r.db.WithContext(ctx).Save(model).Error
}

// Delete removes a page by its ID
func (r *pageRepository) Delete(ctx context.Context, id uuid.UUID) error {
	result := r.db.WithContext(ctx).Delete(&models.Page{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

File: backend/src/database/repository/page.rs
use async_trait::async_trait;
use chrono::Utc;
use diesel::{
    ExpressionMethods, OptionalExtension, QueryDsl, SelectableHelper,
    dsl::count_star,
    result::{DatabaseErrorKind, Error as DieselError},
};
use diesel_async::RunQueryDsl;
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::{
    core::page::{Page, PageListOptions, PageRepository},
    database::{
        models::page::{NewPageRow, PageRow},
        pool::PgPool,
    },
    error::AppError,
    schema::page,
};

#[derive(Clone)]
pub struct DieselPageRepository {
    pool: PgPool,
}

impl DieselPageRepository {
    pub const fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }
}

#[async_trait]
impl PageRepository for DieselPageRepository {
    async fn find_by_id(&self, id: Uuid) -> Result<Page, AppError> {
        let mut connection = self.connection().await?;
        page::table
            .find(id)
            .select(PageRow::as_select())
            .first::<PageRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .ok_or(AppError::NotFound)?
            .try_into()
    }

    async fn find_by_slug(&self, slug: &str) -> Result<Page, AppError> {
        let mut connection = self.connection().await?;
        page::table
            .filter(page::slug.eq(slug))
            .select(PageRow::as_select())
            .first::<PageRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .ok_or(AppError::NotFound)?
            .try_into()
    }

    async fn list(&self, options: PageListOptions) -> Result<(Vec<Page>, i64), AppError> {
        let offset = pagination_offset(options)?;
        let mut connection = self.connection().await?;

        let mut count_query = page::table.into_boxed();
        if let Some(is_published) = options.is_published {
            count_query = count_query.filter(page::is_published.eq(is_published));
        }
        let total = count_query
            .select(count_star())
            .first::<i64>(&mut connection)
            .await
            .map_err(map_diesel_error)?;

        let mut query = page::table.into_boxed();
        if let Some(is_published) = options.is_published {
            query = query.filter(page::is_published.eq(is_published));
        }
        let rows = query
            .order(page::created_at.desc())
            .offset(offset)
            .limit(options.per_page)
            .select(PageRow::as_select())
            .load::<PageRow>(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        let pages = rows
            .into_iter()
            .map(Page::try_from)
            .collect::<Result<Vec<_>, _>>()?;
        Ok((pages, total))
    }

    async fn save(&self, value: &mut Page) -> Result<(), AppError> {
        if value.id.is_nil() {
            value.id = Uuid::new_v4();
        }
        let meta_data = meta_data_to_value(value);
        let mut connection = self.connection().await?;
        let exists = page::table
            .find(value.id)
            .select(page::id)
            .first::<Uuid>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .is_some();

        if exists {
            let row = diesel::update(page::table.find(value.id))
                .set((
                    page::slug.eq(&value.slug),
                    page::title.eq(&value.title),
                    page::content.eq(&value.content),
                    page::description.eq(&value.description),
                    page::image_url.eq(&value.image_url),
                    page::meta_data.eq(meta_data),
                    page::is_published.eq(value.is_published),
                    page::updated_at.eq(Utc::now()),
                ))
                .returning(PageRow::as_returning())
                .get_result::<PageRow>(&mut connection)
                .await
                .map_err(map_diesel_error)?;
            *value = row.try_into()?;
            Ok(())
        } else {
            let row = diesel::insert_into(page::table)
                .values(NewPageRow {
                    id: value.id,
                    slug: &value.slug,
                    title: &value.title,
                    content: &value.content,
                    description: &value.description,
                    image_url: &value.image_url,
                    meta_data,
                    // GORM's `default:true` tag substitutes the database default
                    // when a newly-created model carries bool's false zero value.
                    is_published: value.is_published.then_some(true),
                })
                .returning(PageRow::as_returning())
                .get_result::<PageRow>(&mut connection)
                .await
                .map_err(map_diesel_error)?;
            *value = row.try_into()?;
            Ok(())
        }
    }

    async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::delete(page::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if affected == 0 {
            Err(AppError::NotFound)
        } else {
            Ok(())
        }
    }
}

fn pagination_offset(options: PageListOptions) -> Result<i64, AppError> {
    options
        .page
        .checked_sub(1)
        .and_then(|page| page.checked_mul(options.per_page))
        .ok_or_else(|| AppError::InvalidInput("pagination is out of range".to_owned()))
}

fn meta_data_to_value(page: &Page) -> Option<Value> {
    page.meta_data
        .clone()
        .map(|values| Value::Object(values.into_iter().collect::<Map<_, _>>()))
}

fn map_diesel_error(error: DieselError) -> AppError {
    match error {
        DieselError::NotFound => AppError::NotFound,
        DieselError::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
            AppError::Conflict("resource already exists".to_owned())
        }
        _ => AppError::Database,
    }
}
